### PR TITLE
Fix invalid browser name, version and plugins messages (#3051)

### PR DESCRIPTION
**Pull request by zinduolis**
# Pull Request

Thanks for submitting a PR! Please fill in this template where appropriate:

## Category
Bug

## Feature/Issue Description
**Q:** Please give a brief summary of your feature/fix
**A:** Fix invalid browser version and plugins messages (#3051)

**Q:** Give a technical rundown of what you have changed (if applicable)
**A:** Updated conditions checking for valid browser version. Also added message for situation when plugins are not detected (e. g. for mobile browsers). Added https://github.com/bowser-js/bowser and extended Chrome, Firefox and Opera browser name recognition.

## Test Cases
**Q:** Describe your test cases, what you have covered and if there are any use cases that still need addressing.
**A:** Successfully ran bundle exec rake, had print_info in the browser.rb file to confirm the values when hooking Firefox, Chrome, Opera and Android Chrome. Also, executed all the debug modules with these 3 browser types without any errors on beef side.

## Wiki Page
*If you are adding a new feature that is not easily understood without context, please draft a section to be added to the Wiki below.*

### CODE REVIEW

**Comment by bcoles** [comment]
`window.navigator.userAgent` cannot be trusted. This is why we perform checks for each browser version using browser fingerprinting.

Although the current implementation is insane (it was developed when there was usually years between major releases), we should not throw it out in favor of weak detection (ie, untrusted user agent string).

The entire browser fingerprinting system needs to be reworked.

If your goal is to make the "invalid browser" error message go away, please extend the existing functionality.

**Comment by bcoles** [comment]
If you would prefer to rework the browser fingerprinting, please see:

* #1392

**Comment by zinduolis** [comment]
Thanks, @bcoles, for your feedback, I have extended the existing functionality. 

In scope of #3051, I would like to address the invalid name / version and plugins issues. 

I've noticed that the browser fingerprinting functionality could benefit from significant refactoring. I'd be happy to take on this task once I complete the current fix, given its priority and my availability.

**Comment by zinduolis** [comment]
Added fix for version and renamed the PR.

**Comment by zinduolis** [comment]
Hey @bcoles , @stephenakq , I believe I've resolved the issue. Can we merge this one?

File: core/filters/browser.rb [hunk 1]
```diff
@@ -46,7 +46,7 @@ def self.is_valid_browserversion?(str)
       return false if has_non_printable_char?(str)
       return true if str.eql? 'UNKNOWN'
       return true if str.eql? 'ALL'
-      return false if !nums_only?(str) and !is_valid_float?(str)
+      return false if !nums_only?(str) and !str.match(/\b\d+(\.\d+)+\b/)
```
Comment (bcoles) [inline core/filters/browser.rb:49]: Why is this change necessary? The browser version should not contain `\b`.

**Review by bcoles — COMMENTED** [review]
The user agent cannot be trusted. Please extend the existing approach to support new browser versions.

File: core/filters/browser.rb (hunk 1 above)
Comment (zinduolis) [reply core/filters/browser.rb:49]: Float doesn't take into account longer version numbers separated by dots. FF passes that with number that contains one dot (e. g. 130.4), however, C fails with multiple dots (e. g. 131.0.6738.0).

Without word boundaries, the regex becomes more permissive and can match sequences that are embedded within larger strings, even if that’s not what you want. Word boundaries (\b) are crucial when you need to ensure the version number (or similar pattern) is isolated and not part of a larger string or word.

Comment (bcoles) [reply core/filters/browser.rb:49]: The suggested changes do not preserve the validation logic. These changes allow strings containing `\b` to be treated as valid browser versions.

The core filters are used for validation. This method is used to validate the browser version in multiple areas of the application.

The current implementation first checks for `ALL` and `UNKNOWN` (this is due to poor design). It then rejects any values which are not numbers or floats. While the float check is clearly an oversight as `major.minor.revision` strings should be supported (`/[0-9\.]+/`), the suggested approach will match other values such as `test 1.2.3.4 test` (so long as it is less than 20 characters) which is not a valid browser version.

```ruby
#!/usr/bin/env ruby

def valid?(str)
  str.match(/\b\d+(\.\d+)+\b/) ? true : false
end

puts valid?('test 1.2.3.4 test')

```

You probably want to do something like this:

```suggestion
      return false unless str.match(/\A\d+([\d\.]+)\z/)
```

Although this approach is also flawed in that browser versions can feasibly contain letters. `1.2.3a` or `1.2.3-alpha` for example.

A more robust long term approach would be to leverage the `Gem::Version` library and split the input string by `-` and ` `. However, this is outside the scope of this PR.

Comment (zinduolis) [reply core/filters/browser.rb:49]: Thanks, mate, for elaboration, good point, yours is better than mine due to the beginning and ending anchors rather than word boundaries. However, I've noticed a slight discrepancy in there, it allows for strings like 1.1.1. (ending with the dot). I'm about to commit a slightly modified version of it:  `str.match(\A\d+(\.\d+)+\z)` .

**Comment by zinduolis** [comment]
Hi @bcoles , am I still missing anything or this can finally be merged? Thanks

**Comment by zinduolis** [comment]
@bcoles , I have reverted the UA check to find browser name. Any issues with the fixes for version and plugin checks?

**Review by coreysh12 — APPROVED** [review]

**Comment by zinduolis** [comment]
Hey @stephenakq , as requested I have added https://github.com/bowser-js/bowser into the code and extended the browser name recognition functionality for Chrome, Firefox and Opera. This PR now has everything that's needed to address #3051  

Could you please merge this PR now? Thanks
